issues/347: qfpolkavm's pallet execute ADR (#351)
QuantumFusion-network/spec: docs/ADR/0010_qfpolkavm_pallet_execute.md

Reviewer: Please add `Decision` section.

> ## Title: Design of Polka VM pallet's execute() extrinsics

Reviewer 2: ```suggestion
## Title: Design of Polka VM pallet's execute() extrinsic
```

## Date
2025-05-20
Last status update: 2025-05-21

## Status
- [X] Proposed
- [ ] Accepted
- [ ] Rejected
- [ ] Deprecated
- [ ] Superseded by ADR #X

### Implementation Status
- [ ] Planned
- [ ] In Development
- [ ] Implemented
- [ ] Verified
- [ ] Discontinued

## People
- **Decision-makers**: Andrew, Alisher, Vas
- **Consulted**: KP
- **Informed**:
  - Svat

## Context
For running a smart contract we have to have implemented pallet's extrinsic execute() and at least one method inside the contract to run it (i.e. main()).

## Problem
We have introduced [pallet's execute method](https://github.com/QuantumFusion-network/qf-solochain/blob/main/pallets/qf-polkavm/src/lib.rs#L333), it has signature:
```
execute(origin, contract_address, to, value, user_data, gas_limit, gas_price)
```, where:
- `origin` is a caller address,

> - `contract_address` is an adress of stored onchain contract,

Reviewer 2: ```suggestion
- `contract_address` is an address of stored on-chain contract,
```

> - `to` is an account_id used for balance_transfer() and balance_of(), these are contract functions,

Reviewer 2: ```suggestion
- `to` is an account_id used for transfer() and balance_of(), these are contract functions,
```

> - `value` is a value for balance_transfer(), the contract functions,

Reviewer 2: ```suggestion
- `value` is a value for transfer(), the contract functions,
```

- `user_data` is SCALE encoded structures passed to the contract's main() function,
- `gas_limit` is a limit of the gas for executing of the contract behind contract_address,

> - `gas_price` is a price of the gas for auction; this argument is not Implemented, it's for the future.

Reviewer 2: ```suggestion
- `gas_price` is a price of the gas for auction; this argument is not implemented, it's for the future.
```

As we can see the signature looks at least strage and contains legacy (from PoC) arguments, which are `to`, `value`.

## Decision

> ### Change signature of pallet's extrinsic to:

Author: Since the first implementation was more PoC. Not sure that make sense to add this information.

```
execute(origin, contract_address, data, gas_limit, storage_deposit_limit, gas_price)
```
, where:
- `origin` is a caller address,
- `contract_address` is an address of stored on-chain contract,

> - `data` is [SCALE] encoded structures passed to the contract's context and available via host function call,

Reviewer: If serialization is up to the smart contract developer we should not specify SCALE as the requirement here.

Author: Correct comment, fixed in: d8a3ab7c894b9d6b073d63fb03c85f9fcd6dea82

> - `data` is SCALE encoded structures passed to the contract's main() function,

Reviewer 2: ```suggestion
- `data` is [SCALE] encoded structures passed to the contract's main() function,
```

- `gas_limit` is a limit of the gas for executing of the contract behind contract_address,

> - `storage_deposit_limit` is a limit of how many storage space could be used for storing data during contract execution, if out of this limit, then transaction fail with Out of Storage limit error message and no data stored. Thing about this case: we have a huge `gas_limit` and trying to save inside a contract, by mistaken, a huge blob of data. This argument is not implemented, it's for the future. Unit: bytes per a storage slot.

Reviewer 3: ```suggestion
- `storage_deposit_limit` is a maximum amount of balance that can be charged from the caller to pay for the storage consumed during contract execution. If this limit is exceeded, then transaction fails with `Out of Storage limit` error message and no data stored. Thing about this case: we have a huge `gas_limit` and trying to save inside a contract, by mistaken, a huge blob of data. This argument is not implemented, it's for the future. Unit: balance.
```

- `gas_price` is a price of the gas for auction; this argument is not implemented, it's for the future.

### Update function signatures

> `transfer()` and `balance_of` should read address and value from the `data` instead of `to` and `value`.

Reviewer 3: ```suggestion
`transfer()` and `balance_of()` should read address and value from the `data` instead of `to` and `value`.
```

- Required dependencies
- Migration steps
- Testing considerations]

## References
[Ref to pallet contracts](https://docs.rs/pallet-contracts/latest/pallet_contracts/pallet/struct.Pallet.html#method.call)
[Concern: value](https://docs.rs/pallet-contracts/latest/pallet_contracts/pallet/struct.Pallet.html#method.call)
Research result: seems `value` introduced for optimization puposes for balance transfering and balance_of. I don't find any other reasons.

> [Storage deposit limit explanation](https://github.com/paritytech/substrate-contracts-node/issues/23#issuecomment-1008751578)

Reviewer 2: [SCALE]: https://github.com/paritytech/parity-scale-codec

Author: Done & pushed

## Title: Design of Polka VM pallet's execute() extrinsic

## Date
2025-05-20
Last status update: 2025-05-21

## Status
- [X] Proposed
- [ ] Accepted
- [ ] Rejected
- [ ] Deprecated
- [ ] Superseded by ADR #X

### Implementation Status
- [ ] Planned
- [ ] In Development
- [ ] Implemented
- [ ] Verified
- [ ] Discontinued

> ## People
> - **Decision-makers**: Andrew, Alisher, Vas
> - **Consulted**: KP
> - **Informed**:
>   - Svat

Reviewer: ```suggestion
## People
- **Decision-makers**: KP, Vas
- **Consulted**: Andrew, Alisher
- **Informed**:
  - Svat
```

## Context
For running a smart contract we have to have implemented pallet's extrinsic execute() and at least one method inside the contract to run it (i.e. main()).

## Problem
We have introduced [pallet's execute method](https://github.com/QuantumFusion-network/qf-solochain/blob/main/pallets/qf-polkavm/src/lib.rs#L333), it has signature:
```
execute(origin, contract_address, to, value, user_data, gas_limit, gas_price)
```, where:
- `origin` is a caller address,
- `contract_address` is an address of stored on-chain contract,

> - `to` is an account_id used for transfer() and balance_of(), these are contract functions,

Author: Good catch. Fixed in: 60b22a39ec7c3dabb2c8f1ae83ceb91fc57bad58

- `value` is a value for transfer(), the contract functions,
- `user_data` is [SCALE] encoded structures passed to the contract's main() function,
- `gas_limit` is a limit of the gas for executing of the contract behind contract_address,
- `gas_price` is a price of the gas for auction; this argument is not implemented, it's for the future.

As we can see the signature looks at least strage and contains legacy (from PoC) arguments, which are `to`, `value`.

## Decision
Change signature of pallet's extrinsic to:
```
execute(origin, contract_address, data, gas_limit, storage_deposit_limit, gas_price)
```, where:
- `origin` is a caller address,
- `contract_address` is an address of stored on-chain contract,

> - `data` is [SCALE] encoded structures passed to the contract's main() function,

Reviewer 4: It's available by the host function call. Not in arguments of `main()`

> passed to the contract's main() function

Author: Fixed in: a44acfec61cd0ae4911535cc48e5988419efca44

- `gas_limit` is a limit of the gas for executing of the contract behind contract_address,

> - `storage_deposit_limit` is a limit of how many storage space could be used for storing data during contract execution, if out of this limit, then transaction fail with Out of Storage limit error message and no data stored. Thing about this case: we have a huge `gas_limit` and trying to save inside a contract, by mistaken, a huge blob of data. This argument is not implemented, it's for the future.

Reviewer 4: Should we describe the units here? Bytes / storage slots?

Author: Fixed in a44acfec61cd0ae4911535cc48e5988419efca44

> - `storage_deposit_limit` is a limit of how many storage space could be used for storing data during contract execution, if out of this limit, then transaction fail with Out of Storage limit error message and no data stored. Thing about this case: we have a huge `gas_limit` and trying to save inside a contract, by mistaken, a huge blob of data. This argument is not implemented, it's for the future.

Reviewer 4: Why does it called `storage_deposit_limit`? Maybe `storage_increase_limit` or something else?

Author: Name has been taken from the polkadot's pallet contract. Reason: I guess more understandable for the community. So I don't have any specific ideas here.

> - `gas_price` is a price of the gas for auction; this argument is not implemented, it's for the future.

Author: Got you point, I guess we could talk about this tomorrow, since we have aprroved this signature, if we planing to change it we need to re-approve it.
